src/library/scala/collection/mutable/Map.scala

```diff
@@ -171,17 +171,19 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
     * @param p  The test predicate
     */
   def filterInPlace(p: (K, V) => Boolean): this.type = {
-    if (nonEmpty) {
-      val array = this.toArray[Any] // scala/bug#7269 toArray avoids ConcurrentModificationException
-      val arrayLength = array.length
-      var i = 0
-      while (i < arrayLength) {
-        val (k, v) = array(i).asInstanceOf[(K, V)]
-        if (!p(k, v)) {
-          this -= k
+    if (!isEmpty) this match {
+      case tm: concurrent.TrieMap[_, _] => tm.asInstanceOf[concurrent.TrieMap[K, V]].filterInPlaceImpl(p)
```

Comment:
Why this pattern vs overriding the method on TrieMap?

Reply:
the method returns `this.type`, which is the type of its class in bytecode. thus, the override has a different return type, and it is not binary compatible